From 08802a81c63bd088c4737aad81e0b25e49dd34b5 Mon Sep 17 00:00:00 2001
From: Mengkejiergeli Ba <mengkejiergeli.ba@intel.com>
Date: Fri, 1 Mar 2024 16:12:27 +0800
Subject: [PATCH 21/98] msdk: Fix session close failure

In the case of multi-channels transcoding, a context with child
sesseion can be parent for others, so we need to check if the
msdkcontext has any child session in the list to avoid session
leaks. Otherwise, we will see the failure of closing a parent
session because one of its child's child session not released.

Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/6259>
---
 sys/msdk/gstmsdkcontext.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/sys/msdk/gstmsdkcontext.c b/sys/msdk/gstmsdkcontext.c
index 12330f8d177f..c15496a581e5 100644
--- a/sys/msdk/gstmsdkcontext.c
+++ b/sys/msdk/gstmsdkcontext.c
@@ -376,6 +376,10 @@ gst_msdk_context_finalize (GObject * obj)
 
   /* child sessions will be closed when the parent session is closed */
   if (priv->parent_context) {
+    /* A context with parent_context can also be a parent to others,
+     * and we need to check its child_session_list */
+    if (priv->child_session_list)
+      g_list_free_full (priv->child_session_list, release_child_session);
     gst_object_unref (priv->parent_context);
     goto done;
   } else
-- 
2.43.0

